### PR TITLE
Add unified CEBRA encoder: pytorch implementation

### PR DESCRIPTION
**Pull request by CeliaBenquet**
This PR adds a PyTorch implementation of a unified CEBRA encoder, which is composed of:

- A new sampling scheme that samples across all sessions so that they can be aligned on the neuron axis to train a single encoder.
- A unified Dataset and Loader, adapted to the new sampling scheme.
- A unified Solver that considers multiple sessions to be aligned at inference.
- A new masked modeling training option, with different types of masking.

🚧 A preprint is pending "Unified CEBRA Encoders for Integrating Neural Recordings via Behavioral Alignment" by Célia Benquet, Hossein Mirzaei, Steffen Schneider, Mackenzie W. Mathis. 

💻 A DEMO Notebook is available at: https://cebra.ai/docs/demos.html

Screenshot 2025-06-05 at 14 26 43

### CODE REVIEW

File: cebra/data/single_session.py [hunk 1]
```diff
@@ -64,9 +64,9 @@ def __len__(self):
     def load_batch(self, index: BatchIndex) -> Batch:
         """Return the data at the specified index location."""
         return Batch(
-            positive=self[index.positive],
-            negative=self[index.negative],
-            reference=self[index.reference],
+            positive=self.apply_mask(self[index.positive]),
```
Comment (MMathisLab) [inline cebra/data/single_session.py:67]: quick sanity check; this is backwards compatable? @CeliaBenquet

File: cebra/datasets/__init__.py [hunk 2]
```diff
@@ -97,6 +97,8 @@ def get_datapath(path: str = None) -> str:
     from cebra.datasets.hippocampus import *
     from cebra.datasets.monkey_reaching import *
     from cebra.datasets.synthetic_data import *
+    from cebra.datasets.perich import *
```
Comment (MMathisLab) [inline cebra/datasets/__init__.py:100]: do we need to package this, or is this downloaded from DANDI?

File: cebra/integrations/sklearn/cebra.py [hunk 3]
```diff
@@ -497,6 +497,8 @@ class CEBRA(TransformerMixin, BaseEstimator):
             optimizer documentation in :py:mod:`torch.optim` for further information on how to format the
             arguments.
             |Default:| ``(('betas', (0.9, 0.999)), ('eps', 1e-08), ('weight_decay', 0), ('amsgrad', False))``
+        masking_kwargs (dict):
+            TODO(celia)
```
Comment (MMathisLab) [inline cebra/integrations/sklearn/cebra.py:501]: Let's finalize this :D @CeliaBenquet

File: cebra/models/model.py [hunk 4]
```diff
@@ -676,22 +680,7 @@ def __init__(self, num_neurons, num_units, num_output, normalize=True):
...
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
+            *self._make_layers(num_units, num_layers=16),
```
Comment (MMathisLab) [inline cebra/models/model.py:683]: same as above :)

File: cebra/models/model.py [hunk 5]
```diff
@@ -795,12 +817,7 @@ def __init__(self, num_neurons, num_units, num_output, normalize=True):
         super().__init__(
             nn.Conv1d(num_neurons, num_units, 2),
             nn.GELU(),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
+            *self._make_layers(num_units, num_layers=6),
```
Comment (MMathisLab) [inline cebra/models/model.py:820]: same as above

File: cebra/solver/multi_session.py [hunk 6]
```diff
@@ -80,10 +80,11 @@ def _single_model_inference(self, batch: cebra.data.Batch,
             across the sample dimensions, the output data should be aligned and
             ``batch.index`` should be set to ``None``.
         """
-        batch.to(self.device)
-        ref = torch.stack([model(batch.reference)], dim=0)
-        pos = torch.stack([model(batch.positive)], dim=0)
-        neg = torch.stack([model(batch.negative)], dim=0)
+        ref, pos, neg = self._compute_features(batch, model)
+
+        ref = ref.unsqueeze(0)
```
Comment (MMathisLab) [inline cebra/solver/multi_session.py:85]: why the change here?

File: cebra/solver/single_session.py [hunk 7]
```diff
@@ -137,10 +138,7 @@ def _inference(self, batch: cebra.data.Batch) -> cebra.data.Batch:
             across the sample dimensions, the output data should be aligned and
             ``batch.index`` should be set to ``None``.
         """
-        batch.to(self.device)
-        ref = self.model(batch.reference)
-        pos = self.model(batch.positive)
-        neg = self.model(batch.negative)
+        ref, pos, neg = self._compute_features(batch)
```
Comment (MMathisLab) [inline cebra/solver/single_session.py:141]: why the change here?

File: cebra/models/decoders.py
Comment (MMathisLab) [inline cebra/models/decoders.py:1]: why not have decoders somewhere like integrations? to me model is the encoders only cc @stes

File: cebra/models/decoders.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import torch.nn as nn
```
Comment (MMathisLab) [inline cebra/models/decoders.py:1]: why not have decoders somewhere like integrations? to me model is the encoders only cc @stes

File: cebra/solver/multi_session.py (hunk 6 above)
Comment (CeliaBenquet) [reply cebra/solver/multi_session.py:85]: It was duplicate code in all subclasses.

File: cebra/solver/single_session.py (hunk 7 above)
Comment (CeliaBenquet) [reply cebra/solver/single_session.py:141]: same

File: cebra/models/model.py [hunk 9]
```diff
@@ -266,9 +274,7 @@ def __init__(self, num_neurons, num_units, num_output, normalize=True):
         super().__init__(
             nn.Conv1d(num_neurons, num_units, 2),
             nn.GELU(),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
-            cebra_layers._Skip(nn.Conv1d(num_units, num_units, 3), nn.GELU()),
+            *self._make_layers(num_units, num_layers=3),
```
[the remark this answers is not in the document]
Comment (CeliaBenquet) [reply cebra/models/model.py:277]: should be, it's only internal, to avoid repetitive code.

File: cebra/data/masking.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+class MaskedMixin:
+    """A mixin class for applying masking to data.
+
+    Note:
+        This class is designed to be used as a mixin for other classes.
+        It provides functionality to apply masking to data.
+        The `set_masks` method should be called to set the masking types
+        and their corresponding probabilities.
+    """
+    masks = []  # a list of Mask instances
```
Comment (stes) [inline cebra/data/masking.py:33]: does this need to be public?

File: cebra/data/base.py [hunk 11]
```diff
@@ -36,7 +37,7 @@
 __all__ = ["Dataset", "Loader"]
 
 
-class Dataset(abc.ABC, cebra.io.HasDevice):
+class Dataset(abc.ABC, cebra.io.HasDevice, cebra_data_masking.MaskedMixin):
```
Comment (stes) [inline cebra/data/base.py:40]: Is this Mixin used anywhere else in the codebase?

Comment (CeliaBenquet) [reply cebra/data/base.py:40]: no

File: cebra/data/multi_session.py [hunk 12]
```diff
@@ -95,15 +96,31 @@ def __getitem__(self, args) -> List[Batch]:
...
+            batch = [
+                cebra_data.Batch(
+                    reference=session[index.reference[session_id]],
+                    positive=session[index.positive[session_id]],
+                    negative=session[index.negative[session_id]],
+                    index=index.index,
+                    index_reversed=index.index_reversed,
+                ) for session_id, session in enumerate(self.iter_sessions())
+            ]
+        return batch
```
[the remark this answers is not in the document]
Comment (CeliaBenquet) [reply cebra/data/multi_session.py:123]: This is a backward compatibility check for old models, I don't know if it's worth it no? ideally we don't have it I added after Mackenzie's comment

File: cebra/solver/unified_session.py [hunk 13]
```diff
@@ -0,0 +1,445 @@
...
+
+    _variant_name = "unified-session"
+
+    def parameters(self, session_id: Optional[int] = None):  # same as single
+        """Iterate over all parameters."""
+        for parameter in self.model.parameters():
+            yield parameter
+
+        for parameter in self.criterion.parameters():
+            yield parameter
```
[the remark this answers is not in the document]
Comment (CeliaBenquet) [reply cebra/solver/unified_session.py:51]: I changed so that yes and droped it

File: cebra/solver/unified_session.py [hunk 14]
```diff
@@ -0,0 +1,445 @@
...
+        for batch_start in batch_range:
+            batch_end = min(batch_start + batch_size,
+                            len(dataset.get_session(session_id)))
+
+            if batch_start == batch_range[-2]:  # one before last batch
+                last_start = batch_start
+                continue
+            if batch_start == batch_range[-1]:  # last batch, likely incomplete
+                batch_start = last_start
+                batch_end = len(dataset.get_session(session_id))
```
[the remark this answers is not in the document]
Comment (CeliaBenquet) [reply cebra/solver/unified_session.py:241]: It's not duplicate code, the logic is similar but the concatenation of the batches are done on different objects (the batches in teh loader vs indices in the input to transform so that sampling happens in batch.

Both are tested individually

File: cebra/solver/base.py [hunk 15]
```diff
@@ -407,6 +407,32 @@ def parameters(self, session_id: Optional[int] = None):
         """
         raise NotImplementedError
 
+    def _compute_features(
+        self,
+        batch: cebra.data.Batch,
+        model: Optional[torch.nn.Module] = None
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
```
[the remark this answers is not in the document]
Comment (stes) [reply cebra/solver/base.py:414]: So what is the idea, what does `_inference` do vs. `_compute_features`?

File: cebra/data/single_session.py [hunk 16]
```diff
@@ -63,11 +63,21 @@ def __len__(self):
...
+                negative=self.apply_mask(self[index.negative]),
+                reference=self.apply_mask(self[index.reference]),
+            )
+        else:
+            batch = Batch(
+                positive=self[index.positive],
+                negative=self[index.negative],
+                reference=self[index.reference],
+            )
+        return batch
```
[the remark this answers is not in the document]
Comment (CeliaBenquet) [reply cebra/data/single_session.py:80]: This is a backward compatibility check for old models, I don't know if it's worth it no? ideally we don't have it I added after Mackenzie's comment

File: cebra/solver/unified_session.py [hunk 17]
```diff
@@ -0,0 +1,445 @@
...
+        elif padding_mode == "random":
+            padded_inputs = torch.normal(
+                torch.zeros((zero_shape, self.n_features)),
+                torch.ones((zero_shape, self.n_features)))
+
+        else:
+            raise ValueError(
+                f"Invalid padding mode: {padding_mode}. "
+                "Choose from 'time', 'zero', 'poisson', 'random', or 'random_poisson'."
+            )
```
[the remark this answers is not in the document]
Comment (CeliaBenquet) [reply cebra/solver/unified_session.py:342]: If you want to reproduce the experiment, you need that code, but else, yes we can drop if that's better

[the remark this answers is not in the document]
Comment (CeliaBenquet) [reply cebra/solver/unified_session.py:342]: I moved it to an experiment file I have instead, and droped.

File: cebra/integrations/sklearn/cebra.py [hunk 18]
```diff
@@ -497,6 +497,9 @@ class CEBRA(TransformerMixin, BaseEstimator):
             optimizer documentation in :py:mod:`torch.optim` for further information on how to format the
             arguments.
             |Default:| ``(('betas', (0.9, 0.999)), ('eps', 1e-08), ('weight_decay', 0), ('amsgrad', False))``
+        masking_kwargs (dict):
+            A dictionary of masking types and their corresponding required masking values. The keys are the
+            names of the Mask instances.
```
[the remark this answers is not in the document]
Comment (CeliaBenquet) [reply cebra/integrations/sklearn/cebra.py]: done

File: cebra/solver/single_session.py [hunk 19]
```diff
@@ -139,8 +120,8 @@ def _inference(self, batch: cebra.data.Batch) -> cebra.data.Batch:
         """
         batch.to(self.device)
         ref = self.model(batch.reference)
-        pos = self.model(batch.positive)
         neg = self.model(batch.negative)
+        pos = self.model(batch.positive)
```
Comment (stes) [inline cebra/solver/single_session.py:124]:
```suggestion
        pos = self.model(batch.positive)
        neg = self.model(batch.negative)
```

**Review by stes — APPROVED** [review]
Remaining issues are:

- more tests (e.g. for changes to model, adaptations of sklearn API, ...)
- improved class design for masking
- improved class design for solver hierarchy

However, this can be done in follow-up contributions, this can be merged now. Attempt to clean up the commit message for sqashing:

```
* start tests

* remove print statements

* first passing test

* move functionality to base file in solver and separate in functions

* add test_select_model for multisession

* remove float16

* Improve modularity remove duplicate code and todos

* Add tests to solver

* Fix save/load

* Fix extra docs errors

* Add review updates

* apply ruff auto-fixes

* fix linting errors

* Run isort, ruff, yapf

* Fix gaussian mixture dataset import

* Fix all tests but xcebra tests

* Fix pytorch API usage example

* Make xCEBRA compatible with the batched inference & padding in solver

* Add some tests on transform() with xCEBRA

* Add some docstrings and typings and clean unnecessary changes

* Implement review comments

* Fix sklearn test

* Initial pass at integrating unifiedCEBRA

* Add name in NOTE

* Implement reviews on tests and typing

* Fix import errors

* Add select_model to aux solvers

* Fix tests

* Add mask tests

* Fix docs error

* Remove masking init()

* Remove shuffled neurons in unified dataset

* Remove extra datasets

* Add tests on the private functions in base solver

* Update tests and duplicate code based on review

* Fix quantized_embedding_norm undefined when `normalize=False` (#249)

* Fix tests

* Adapt unified code to get_model method

* Update mask.py

add headers to new files

* Update masking.py

- header

* Update test_data_masking.py

- header

* Implement review comments and fix typos

* Fix docs errors

* Remove np.int typing error

* Fix docstring warning

* Fix indentation docstrings

* Implement review comments

* Fix circular import and abstract method

* Add maskedmixin to __all__

* Implement extra review comments

* Change masking kwargs as tuple and not dict in sklearn impl

* Add integrations/decoders.py

* Fix typo

* minor simplification in solver

---------

Note, some comments in this PR overlap with
https://github.com/AdaptiveMotorControlLab/CEBRA/pull/168
and
https://github.com/AdaptiveMotorControlLab/CEBRA/pull/225
which were developed in parallel.
```
